### PR TITLE
Let JobHost.Call pass route parameters when invoking. See Unit test.

### PR DESCRIPTION
**Pull request by MikeStall**

### CODE REVIEW

File: test/Microsoft.Azure.WebJobs.Host.UnitTests/Common/HostCallTestsWithRouteData.cs [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+            activator.Add(testInstance);
+
+            IExtensionRegistry extensions = config.GetService<IExtensionRegistry>();
+            FakeExtClient client = new FakeExtClient();
+            extensions.RegisterExtension<IExtensionConfigProvider>(client);
+
+            JobHost host = new JobHost(config);
+
+            var method = typeof(Functions).GetMethod("Func");
+            await host.CallAsync(method, new { k1 = 100, k2 = 200 });
```
Comment (mathewc) [inline test/Microsoft.Azure.WebJobs.Host.UnitTests/Common/HostCallTestsWithRouteData.cs]: The "arguments" parameter name has xml doc indicating that the keys are parameter names of the target method. The values you're passing in here are not method arguments.

File: src/Microsoft.Azure.WebJobs.Host/Bindings/FunctionBinding.cs [hunk 2]
```diff
@@ -25,7 +25,9 @@ public FunctionBinding(FunctionDescriptor descriptor, IReadOnlyDictionary<string
         {
             Dictionary<string, IValueProvider> results = new Dictionary<string, IValueProvider>();
 
-            BindingContext bindingContext = new BindingContext(context, null);
+            // Supplied bindings can be direct parameters or route parameters.
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Host/Bindings/FunctionBinding.cs]: FunctionBinding is used for functions w/o trigger bindings only. Doesn't that mean that what you're trying to do here won't work for triggered functions?

File: src/Microsoft.Azure.WebJobs.Host/Bindings/FunctionBinding.cs [hunk 3]
```diff
@@ -25,7 +25,9 @@ public FunctionBinding(FunctionDescriptor descriptor, IReadOnlyDictionary<string
         {
             Dictionary<string, IValueProvider> results = new Dictionary<string, IValueProvider>();
 
-            BindingContext bindingContext = new BindingContext(context, null);
+            // Supplied bindings can be direct parameters or route parameters. 
+            IReadOnlyDictionary<string, object> routeParameters = (parameters != null) ? new Dictionary<string, object>(parameters) : null;
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Host/Bindings/FunctionBinding.cs]: nit: rename routeParameters to bindingData

File: src/Microsoft.Azure.WebJobs.Host/Bindings/FunctionBinding.cs (hunk 2 above)
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Host/Bindings/FunctionBinding.cs]: Also, I'm a little concerned about overloading this dictionary so that it contains both raw argument values (the current behavior) as well as binding data (the new behavior you're adding).

Comment (MikeStall) [inline src/Microsoft.Azure.WebJobs.Host/Bindings/FunctionBinding.cs]: Correct - this won't solve trigger functions. 

---
In reply to: [62585581](https://github.com/Azure/azure-webjobs-sdk/pull/719#discussion_r62585581) [](ancestors = 62585581)

File: src/Microsoft.Azure.WebJobs.Host/Bindings/FunctionBinding.cs (hunk 3 above)
Comment (MikeStall) [inline src/Microsoft.Azure.WebJobs.Host/Bindings/FunctionBinding.cs]: ok

---
In reply to: [62585863](https://github.com/Azure/azure-webjobs-sdk/pull/719#discussion_r62585863) [](ancestors = 62585863)

File: test/Microsoft.Azure.WebJobs.Host.UnitTests/Common/HostCallTestsWithRouteData.cs (hunk 1 above)
Comment (MikeStall) [inline test/Microsoft.Azure.WebJobs.Host.UnitTests/Common/HostCallTestsWithRouteData.cs]: This is the goal of the change.   I can update the xml docs. 

---
In reply to: [62585328](https://github.com/Azure/azure-webjobs-sdk/pull/719#discussion_r62585328) [](ancestors = 62585328)

File: src/Microsoft.Azure.WebJobs.Host/JobHost.cs [hunk 4]
```diff
@@ -194,7 +194,7 @@ public void Call(MethodInfo method)
         /// <param name="method">The job method to call.</param>
         /// <param name="arguments">
         /// An object with public properties representing argument names and values to bind to parameters in the job
-        /// method.
+        /// method. These may be direct parameter values of bindingData.
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Host/JobHost.cs]: nit: can you reword this to "In addition to parameter values, these may also include binding data values".

File: src/Microsoft.Azure.WebJobs.Host/JobHost.cs [hunk 5]
```diff
@@ -203,7 +203,8 @@ public void Call(MethodInfo method, object arguments)
 
         /// <summary>Calls a job method.</summary>
         /// <param name="method">The job method to call.</param>
-        /// <param name="arguments">The argument names and values to bind to parameters in the job method.</param>
+        /// <param name="arguments">The argument names and values to bind to parameters in the job method.
+        /// These may be direct parameter values of bindingData. </param>
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Host/JobHost.cs]: Same here and below.

File: test/Microsoft.Azure.WebJobs.Host.UnitTests/Common/HostCallTestsWithRouteData.cs [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace Microsoft.Azure.WebJobs.Host.UnitTests
+{
+    // Unit test for exercising Host.Call passing route data.
```
Comment (mathewc) [inline test/Microsoft.Azure.WebJobs.Host.UnitTests/Common/HostCallTestsWithRouteData.cs]: nit: I'd also replace all this "route data" terminology and file name with "binding data". There are no routes here

**Comment by mathewc** [comment]
A few final comments, otherwise looks good. :ship:

File: src/Microsoft.Azure.WebJobs.Host/JobHost.cs (hunk 4 above)
Comment (MikeStall) [inline src/Microsoft.Azure.WebJobs.Host/JobHost.cs]: yes.

---
In reply to: [62775546](https://github.com/Azure/azure-webjobs-sdk/pull/719#discussion_r62775546) [](ancestors = 62775546)

**Comment by MikeStall** [comment]
Fixes #721 . Pushed to master.
